ovn-org/ovn-kubernetes
# Utilize HW offloading for uplink <--> ext.bridge traffic by using bypass port

**- What this PR does and why is it needed**
HW capable devices, like Mellanox ConnectX or Bluefield2 cards, can't fully offload CT flows where bridge is in/out port, because CT infrastructure by design can't provide all information needed to apply HW offloading.
Following PR enables HW offloading for uplink <--> ext.bridge traffic by using bypass port effectively replacing bridge as in/out port, so the traffic, in fact, goes between uplink and bypass port. To use this feature user must specify [switchdev VF](https://www.kernel.org/doc/html/latest/networking/switchdev.html) (don't confuse with SRIOV VF) or [Subfunction](https://github.com/Mellanox/scalablefunctions/wiki) as a `--gateway-interface` or pre-create bridge, plug VF/SF representor to the bridge, and set node IP on VF/SF. In both cases worker node must end up with following bridge layout:

```
                        ┌──────────┐
                        │  br-ext  │
                  ┌─────┴──┐       │    ┌──────────┐
                  │  eth0  │       │    │  br-int  │
                  └─────┬──┘       │    │          │
                        │          X────X          │
   ┌────────┐     ┌─────┴──┐       │    │          │
   │ eth0v0 ├─────┤ eth0_0 │       │    │          │
   └────────┘     └─────┬──┘       │    └──────────┘
     NODE_IP            │          │
                        └──────────┘
```

Where `eth0v0` and `eth0_0` are, for ex., VF and VF representor of `eth0` uplink.
If user specify SRIOV VF, PF, representor or any other device, except the ones from above, then ovnkube-node will create regular bridge layout as it is now.
Additionally, used netdevice must be excluded from device plugin pools, so it won't be used for workload pods.

**- Special notes for reviewers**
Besides of actual feature commit PR include commit to improve gateway codebase: code reorganized to reduce number of arguments carried through the functions, removes redundant code, etc.. Also unit tests updated and added new test cases.


**- How to verify it**
On the "fresh" node:
1. Create switchdev VF or SF.
2. Assign node IP to uplink and create default route.
3. Deploy a cluster, with ovnkube-node.yaml having `OVN_GATEWAY_OPTS` value of VF or SF netdevice. For the scheme above it is `--gateway-interface=eth0v0`.
4. Deploy some workload pod to see if cluster work as expected.

**- Description for the changelog**
Extend gateway code to support switchdev VF or SF as a gateway interface to utilize HW offloading.
